File: iteration2/code/src/coursetracking/models/Student.java
package coursetracking.models;

import java.util.ArrayList;
import java.util.Collections;
import java.io.File;
import java.io.FileWriter;

import com.google.gson.Gson;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import coursetracking.utils.Utils;

public class Student {

    @SerializedName("id")
    @Expose
    private int id;

    @SerializedName(value = "name", alternate = { "studentName" })
    @Expose
    private String name;

    @SerializedName(value = "surname", alternate = { "studentSurname" })
    @Expose
    private String surname;

    @SerializedName("transcripts")
    @Expose
    private ArrayList<Transcript> transcripts;

    @SerializedName("currentTranscript")
    @Expose
    private Transcript currentTranscript;

    @SerializedName("gpa")
    @Expose
    private float gpa;

    @SerializedName("semester")
    @Expose
    private int semester;

    @SerializedName("currentCourses")
    @Expose
    private ArrayList<Course> currentCourses;
    
    @SerializedName("passedCourses")
    @Expose
    private ArrayList<String> passedCourses;

    @SerializedName("feedback")
    @Expose
    private ArrayList<String> feedback;

    @SerializedName("totalCredit")
    @Expose
    private int totalCredit;

    public Student() {
        // empty constructor.
    }

    public Student(Student student) {
        this.id = student.id;
        this.name = student.name;
        this.surname = student.surname;
        this.semester = student.semester;
        this.currentCourses = null;
        this.transcripts = null;
    }

    public Student(int id, String name, String surname, int semester) {
        this.id = id;
        this.name = name;
        this.surname = surname;
        this.semester = semester;
        this.currentCourses = null;
        this.transcripts = null;
    }

    public void calculate() {
        sortTranscripts();
        calculatePassedCourses();
        calculateTotalCredit();
        calculateSemester();
    }

    public boolean addCourse(Course course) {
        if (currentCourses == null)
            currentCourses = new ArrayList<>();
        if (currentCourses.contains(course))
            return false;
        if (!canTakeCourse(course))
            return false;
        currentCourses.add(course);
        return true;
    }

    public boolean register() {
        Transcript tr = new Transcript();
        for (Course c : currentCourses)
            tr.addCourse(new TakenCourse(c, null));
        this.currentTranscript = tr;
        this.save();
        return true;
    }

    /* checks whether the student passed prerequisites of the course
     and adds feedback into student's json if necessary */
    public boolean canTakeCourse(Course course) {
    	if (feedback == null) feedback = new ArrayList<>();	
    	if (course.getPrerequisites() != null) {
    		
    	nextpr: for (Course pr : course.getPrerequisites()) {
    	    		for (Transcript tr : transcripts) {
    					for (TakenCourse c : tr.getCourses()) {
    						if ((c.getCourseCode().equals(pr.getCourseCode()) && c.getLetterGrade().compareTo("DD") <= 0)) {
    							continue nextpr; // switches other prereq. if there is
    						}
    					}
    				}
    				// if the prereq. isn't passed adds a related feedback and returns false
    	    		if(course.prereqProblemStd == null) course.prereqProblemStd = new ArrayList<>();
    	    		course.prereqProblemStd.add(this);
    				feedback.add("The system did not allow " + course.courseCode + " because student failed prereq. " + pr.courseCode);
    				return false; 
    			}
    			return true; // if there is no prereq. problem returns true
    		}
    	return true; // if the course has no prereq. returns true 	
    }

    public void addFeedbackQuota(Course course) {
    	if (feedback == null) feedback = new ArrayList<>();
        feedback.add("The student could not register for "+ course.getCourseCode() +" because of a quota problem");	
    }

    public void calculateSemester() {
        float cumulative = 0f;
        int cumulativeCourseCredit = 0;
        for (Transcript t : transcripts) {
            cumulative += t.getSemesterGPA() * t.getTotalCredit();
            cumulativeCourseCredit += t.getTotalCredit();
            if (cumulative / cumulativeCourseCredit < 1.8f)
                break;
            semester = t.getSemester();
        }
    }

    public void calculateTotalCredit() {
        int sumOfCredits = 0;
        for (Transcript t : transcripts) {
            for (TakenCourse c : t.getCourses()) {
                if (Utils.getInstance().getGPAOfLetterGrade(c.getLetterGrade()) > 0.1f)
                    sumOfCredits += c.getCredit();
            }
        }
        totalCredit = sumOfCredits;
    }

    public void calculatePassedCourses() {
        if (passedCourses == null)
            passedCourses = new ArrayList<>();
        for (Transcript t : transcripts) {
            for (TakenCourse c : t.getCourses()) {
                if (Utils.getInstance().getGPAOfLetterGrade(c.getLetterGrade()) - 0.01 > Utils.getInstance()
                        .getGPAOfLetterGrade("FF") && !passedCourses.contains(c.courseCode)) {
                    passedCourses.add(c.courseCode);
                }
            }
        }
    }
  
    public boolean isStudentEnrolled(Course c){
        if(currentCourses == null) return false;
        for(Course currentCourse: this.currentCourses){
            if(c.getCourseCode() == currentCourse.getCourseCode())
            return true;
        }
        return false;
    }

    public boolean readTranscript() {
        return true;
    }

    public int getTotalCredit() {
        return totalCredit;
    }

    public void sortTranscripts() {
        Collections.sort(transcripts);
    }

    public ArrayList<Transcript> getTranscripts() {
        return transcripts;
    }

    public void calculateCumutlativeGPA(int semester) {// this will implement at register course (transcript) part.
        this.gpa = 4.0f;
    }

    public float getGPA() {
        return gpa;
    }

    public ArrayList<Course> getCurrentCourses() {
        return currentCourses;
    }

    public Transcript getCurrenTranscript() {
        return currentTranscript;
    }

    public void setCurrentTrancript(Transcript currentTranscript) {
        this.currentTranscript = currentTranscript;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public void save() {
        Gson gson = new Gson();

        try {
            File std = new File(Utils.getInstance().getOutputPath(), this.id + ".json");
            std.createNewFile();
            FileWriter writer = new FileWriter(std);
            gson.toJson(this, writer);
            writer.flush();
            writer.close();
        } catch (Exception e) {
            System.out.println(e);
        }

    }

    public void addTranscript(Transcript tr) {
        if (this.transcripts == null)
            this.transcripts = new ArrayList<>();
        this.transcripts.add(tr);
    }
}

File: iteration3/code/src/model/student.py
from typing import List
from src.model.course import Course
from src.model.transcript import Transcript
from src.utils.utils import gpa_letter_to_grade

class Student(object):

    def __init__(self):
        pass

    def __init__(self, _id: int, name: str, surname: str, semester: int):
        self.__id: int = _id
        self.__name: str = name
        self.__surname: str = surname
        self.__semester: int = semester
        self.__currentCourses: List[Course] = []
        self.__transcripts: List[Transcript] = []

        self.__passedCourses: List[str] = []
        self.__feedback: List[str] = []
        self.__totalCredit: int = 0
        self.__currentTranscript: Transcript = None

    def calculate(self):
        self.sortTranscript()
        self.calculatePassedCourses()
        self.calculateTotalCredit()
        self.calculateSemester()

    def addCourse(self, course: Course) -> bool:
        if(self.__currentCourses is None):
            self.__currentCourses = []

        if(course in self.__currentCourses):
            return False

        if(not self.canTakeCourse(course)):
            return False

        self.__currentCourses.append(course)
        return True

    """checks whether the student passed prerequisites of the course
     and adds feedback into student's json if necessary"""

    def canTakeCourse(self, course: Course) -> bool:
        if(course.getPrerequisites()):
            nextpr = False
            for pr in List[Course](course.getPrerequisites()):
                for tr in self.__transcripts:
                    for c in tr.getCourses():
                        # equals() -> == , compareTo -> >= , are we check is letter grade bigger or smaller than DD
                        if((c.getCourseCode() == pr.getCourseCode and c.letterGrade() >= 'DD') <= 0):
                            nextpr = True
                            break
                    if nextpr:
                        break
                if nextpr:
                    continue
                
                # if the prereq. isn't passed adds a related feedback and returns false
                course.addPrereqProblem(self)
                self.addFeedbackPrereq(course, pr)

                return False  # if there is no prereq. problem returns true
            return True  # if the course has no prereq. returns true

    def calculateSemester(self):
        cumulative: float = 0.0
        cumulativeCourseCredit: int = 0

        for t in self.__transcripts:
            cumulative += t.getSemesterGPA() * t.getTotalCredit()
            cumulativeCourseCredit += t.getTotalCredit()

            if(cumulative / cumulativeCourseCredit < 1.8):
                break
            self.__semester = t.getSemester()

    def calculateTotalCredit(self):
        sumOfCredits: int = 0

        for t in self.__transcripts:
            for c in t.getCourses():
                if(gpa_letter_to_grade.get(c.getLetterGrade(), 0.0) > 0.1):
                    sumOfCredits += c.getCredit()

        self.__totalCredit = sumOfCredits

    def calculatePassedCourses(self):
        if(self.__passedCourses is None):
            self.__passedCourses = []
        for t in self.__transcripts:
            for c in t.getCourses():
                if(gpa_letter_to_grade.get(c.getLetterGrade(), 0.0) - 0.01 > gpa_letter_to_grade.get("FF", 0.0) and c.getCourseCode() not in self.__passedCourses):
                    self.__passedCourses.append(c.getCourseCode())

    def isStudentEnrolled(self, c: Course) -> bool:
        if(self.__currentCourses is None):
            return False
        for currentCourse in self.__currentCourses:
            if(c.getCourseCode() == currentCourse.getCourseCode()):
                return True
        return False

    def addFeedback(self, feedback: str):
        if self.__feedback is None:
            self.__feedback = []
        self.__feedback.append(feedback)

    def addFeedbackQuota(self, course: Course) -> None:
        self.addFeedback("The student could not register for " + course.getCourseCode() + " because of a quota problem")

    def addFeedbackPrereq(self, course: Course, prereq: Course):
        self.addFeedback("The system did not allow " + course.getCourseCode() + " because student failed prereq. " + prereq.getCourseCode())

    def setCurrentTranscript(self, currentTranscript: Transcript):
        self.__currentTranscript = currentTranscript

    def setId(self, _id: int):
        self.__id = _id

    def getTotalCredit(self) -> int:
        return self.__totalCredit

    def sortTranscripts(self) -> bool:
        sorted(self.__transcripts)

    def getTranscripts(self) -> List[Transcript]:
        return self.__transcripts

    def getGPA(self) -> float:
        return self.__gpa

    def getCurrentCourses(self) -> List[Course]:
        return self.__currentCourses

    def currentTranscript(self) -> Transcript:
        return self.__currentTranscript

    def getId(self) -> int:
        return self.__id

    def save():

        # TODO read Json file

        pass

    def addTranscript(self, tr: Transcript) -> None:
        if(self.__transcripts is None):
            self.__transcripts = []
        self.__transcripts.append(tr)

    def __str__(self):
        return f"Student({self.stdid}, {self.__name}, {self.__surname}, {self.__semester})"
